tsdb/shard.go

```diff
@@ -43,6 +43,25 @@ var (
 	ErrFieldUnmappedID = errors.New("field ID not mapped")
 )
 
+// A ShardError implements the error interface, and contains extra
+// context about the shard that generated the error.
+type ShardError struct {
+	id  uint64
+	Err error
+}
+
+// NewShardError returns a new ShardError.
+func NewShardError(id uint64, err error) ShardError {
+	if err == nil {
+		err = errors.New("unknown error")
```

Comment:
I can't think of a time when I would want an error value that is non-nil when the actual error that caused it was `nil`. I see that currently this only ever called once, after checking that the error is non-nil. If you instead made `NewShardError` return `nil` when passed a `nil` error value, you could remove the conditional.

More importantly, I think, is that the `nil` error value has always meant that no error occurred. Returning a non-nil error value for a `nil` error would be the same as having `var ErrOK = errors.New("Everything worked great")`, then checking to see if we got the `ErrOK` value to represent that nothing went wrong. We already have a great feature to detect when there wasn't an error: the `nil` error value.